File: 226_Invert_Binary_Tree.java
/*
https://leetcode.com/problems/invert-binary-tree/
Invert a binary tree.

Invert left and right subtrees then swap.
Time - O(n)
Space - O(1)
*/

/**
 * Definition for a binary tree node.
 */

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
    public TreeNode invertTree(TreeNode root) {

        if (root == null)
            return null;

        TreeNode left = invertTree(root.left);
        TreeNode right = invertTree(root.right);

        root.left = right;
        root.right = left;
        return root;
    }
}

File: 226_Invert_Binary_Tree.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/invert-binary-tree/
# Invert a binary tree.

# Left substree is old right subtree inverted. Right subtree is old left subtree inverted.
# Time - O(n)
# Space - O(n)

class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if not root:
            return None

        root.left, root.right = self.invertTree(root.right), self.invertTree(root.left)

        return root
